llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -688,6 +688,15 @@ int main(int argc, char **argv) {
     Err.print(argv[0], errs());
     return 1;
   }
+
+  // Special "llvm.used" variable which holds references to global values in the
+  // module is known to cause problems for tools which run later in pipeline, so
```

Comment:
@sndmitriev, could you clarify what kind of problems might be caused, please?

Reply:
After linking device bitcode llvm.used holds references to the kernels that are defined in the device image. But after splitting device image into separate kernels (with -fsycl-device-code-split=per_kernel option) we may end up with having references to kernel declaration originating from “llvm.used” in the IR that is passed to llvm-spirv tool, and these declarations cause an assertion in llvm-spirv

$ cat test.ll
target triple = "spir64-unknown-unknown-sycldevice"

@llvm.used = appending global [2 x i8*] [i8* bitcast (void ()* @foo to i8*), i8* bitcast (void ()* @bar to i8*)], section "llvm.metadata"

declare spir_kernel void @foo()

define spir_kernel void @bar() {
  ret void
}
$ llvm-as -o test.bc test.ll
$ llvm-spirv -o test.spv -spirv-max-version=1.1 -spirv-debug-info-version=legacy -spirv-allow-extra-diexpressions -spirv-ext=+all,-SPV_INTEL_usm_storage_classes test.bc
llvm-spirv: …/llvm/llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:3438: void SPIRV::LLVMToSPIRV::transFPContract(): Assertion `FPC != FPContract::UNDEF' failed.
…

Reply:
I'm a bit confused with the reasoning.

That FP-contract related assertion seems like an unhandled situation in the translator which should be fixed there.

I do agree that `@llvm.used` might cause problems with the translation, because it contains function pointers and can't be translated into SPIR-V without corresponding extension (`SPV_INTEL_function_pointers`) enabled

Reply:
I do not argue with that, the problem needs to be fixed in the translator as well.

Reply:
How do we get this metadata in device code? It's not emitted in SYCL mode: https://github.com/intel/llvm/commit/aa35161c15b9f9bac661b6084c7ac68f4013fdf4. Is it emitted in other modes?
If so, shouldn't we drop `@llvm.used` during FE IR generation?

Anyway, it would be nice if we add clarification with reasoning to drop `@llvm.used` in sycl-post-link tool from this comment: https://github.com/intel/llvm/pull/2995#discussion_r553112761.

Reply:
> How do we get this metadata in device code?

"llvm.used" variable may appear in the IR after linking with "deps" bitcode file (please see https://github.com/intel/llvm/pull/2872 for more details).

> Anyway, it would be nice if we add clarification with reasoning to drop @llvm.used in sycl-post-link tool from this comment: #2995 (comment).

I have created https://github.com/intel/llvm/pull/3012 to address your comment. Please review it.